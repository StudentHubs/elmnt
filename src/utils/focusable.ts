import * as React from 'react';

export default (setElemName: string, focusName?: string) => C =>
  class Focusable extends React.Component<any> {
    focusElem = null as null | HTMLElement;
    focusWait = false;
    setFocusElem = elem => {
      this.focusElem = elem;
      if (this.focusElem && this.focusWait) {
        this.focusElem.focus();
        this.focusWait = false;
      }
      if (this.props[setElemName]) this.props[setElemName](elem);
    };
    doFocus = event => {
      if (this.focusElem && (event && event.target) !== this.focusElem) {
        this.focusElem.focus();
        event && event.preventDefault();
      }
      if (this.props[focusName!]) this.props[focusName!](event);
    };
    focus = () =>
      this.focusElem ? this.focusElem.focus() : (this.focusWait = true);
    blur = () => this.focusElem && this.focusElem.blur();
    render() {
      return React.createElement(C, {
        ...this.props,
        [setElemName]: this.setFocusElem,
        ...(focusName ? { [focusName]: this.doFocus } : {}),
      });
    }
  } as React.ComponentClass<any>;
