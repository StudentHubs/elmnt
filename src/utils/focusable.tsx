import * as React from 'react';

import { Comp } from '../typings';

export interface FocusableProps {
  focusElem: any;
  setFocusElem: (c: any) => void;
}
interface FocusableState {
  focusElem: HTMLElement | null;
  focusWait: boolean;
}

export default function focusable<P>(InnerComponent: Comp<P & FocusableProps>) {
  return class Focusable extends React.Component<P, FocusableState> {

    public state = { focusElem: null, focusWait: false } as FocusableState;

    private setFocusElem = c => this.setState({ focusElem: c });

    public focus = () => (
      this.state.focusElem ? this.state.focusElem.focus() : this.setState({ focusWait: true })
    );

    public focusElem = () => this.state.focusElem;

    public componentDidUpdate() {
      if (this.state.focusElem && this.state.focusWait) {
        this.state.focusElem.focus();
        this.setState({ focusWait: false });
      }
    }

    public render() {
      return (
        <InnerComponent
          {...this.props} focusElem={this.state.focusElem} setFocusElem={this.setFocusElem}
        />
      );
    }
  } as React.ComponentClass<P>;
}
