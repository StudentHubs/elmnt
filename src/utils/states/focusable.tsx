import * as React from 'react';

import { Comp } from '../../typings';

export interface FocusableProps {
  focusElem: any;
  setFocusElem: (c: any) => void;
}
export interface FocusableOuterProps {
  focusRef?: (c: any) => void;
}
interface FocusableState {
  focusElem: HTMLElement | null;
  focusWait: boolean;
}

export default function focusable<P extends FocusableOuterProps>(
  InnerComponent: Comp<P & FocusableProps>,
) {
  return class Focusable extends React.Component<P, FocusableState> {

    public state = { focusElem: null, focusWait: false } as FocusableState;

    private setFocusElem = c => {
      this.setState({ focusElem: c });
      if (this.props.focusRef) (this.props.focusRef as any)(c);
    }

    public focus = () => (
      this.state.focusElem ? this.state.focusElem.focus() : this.setState({ focusWait: true })
    );

    public isFocusElem = (elem: any) => elem === this.state.focusElem;

    public componentDidUpdate() {
      if (this.state.focusElem && this.state.focusWait) {
        this.state.focusElem.focus();
        this.setState({ focusWait: false });
      }
    }

    public render() {
      return (
        <InnerComponent
          {...this.props} focusElem={this.state.focusElem} setFocusElem={this.setFocusElem}
        />
      );
    }
  } as React.ComponentClass<P>;
}
