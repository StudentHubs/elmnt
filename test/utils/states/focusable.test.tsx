import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import focusable, { FocusableProps } from '../../../src/utils/states/focusable';

describe('utils: focusable', () => {

  it('basic', () => {

    const Comp = focusable<{ style: React.CSSProperties }>(({ setFocusElem, style }) =>
      <div tabIndex={0} ref={setFocusElem} style={style} />
    );

    const tree = renderer.create(<Comp style={{ color: 'red' }} />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<Comp style={{ color: 'red' }} />);
    (wrapper.instance() as any).focus();
    expect(wrapper.find('div').get(0)).toBe(document.activeElement);

    document.body.focus();

  });

  it('focus before elem ready', () => {

    const Comp = focusable<{ show?: boolean }>(({ setFocusElem, show }) =>
      show ? <div tabIndex={0} ref={setFocusElem} /> : null
    );

    const wrapper = mount(<Comp />);
    (wrapper.instance() as any).focus();
    expect(wrapper.find('div').get(0)).toBeUndefined();
    wrapper.setProps({ show: true });
    expect(wrapper.find('div').get(0)).toBe(document.activeElement);

    document.body.focus();

  });

  it('focus on mount', () => {

    const Comp = focusable<{}>(class Inner extends React.Component<FocusableProps, {}> {
      public componentDidUpdate(prevProps) {
        if (this.props.focusElem && !prevProps.focusElem) {
          this.props.focusElem.focus();
        }
      }
      public render() {
        return <div tabIndex={0} ref={this.props.setFocusElem} />;
      }
    });

    const wrapper = mount(<Comp />);
    expect(wrapper.find('div').get(0)).toBe(document.activeElement);

    document.body.focus();

  });

});
