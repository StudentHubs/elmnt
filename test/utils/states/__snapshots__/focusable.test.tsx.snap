// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils: focusable basic 1`] = `
<div
  style={
    Object {
      "color": "red",
    }
  }
  tabIndex={0}
/>
`;
